core/src/test/java/org/jdeferred/impl/ExceptionsTest.java

```diff
@@ -0,0 +1,145 @@
+
+        }
+    }
+
+    @Test
+    public void testFailureExceptionsAreCaught() throws InterruptedException {
+        final DeferredObject<String, Void, Void> deferredObject =
+            new DeferredObject<String, Void, Void>();
+        Promise<String, Void, Void> promise = deferredObject.promise()
+            .except(new ExceptCallback() {
+                @Override
+                public void onException(Exception e) {
+                    assertEquals(THIS_IS_MY_RUNTIME_EXCEPTION, e.getMessage());
+                }
+            })
+            .fail(new FailCallback<Void>() {
+                @Override
+                public void onFail(Void result) {
+                    throw new RuntimeException(THIS_IS_MY_RUNTIME_EXCEPTION);
+                }
+            });
+        deferredObject.reject(null);
+
+        try {
+            promise.waitSafely();
+            fail();
+        } catch (RuntimeException re) {
+        }
+    }
+
+    @Test
+    public void testAlwaysExceptionsAreCaught() throws InterruptedException {
+        final DeferredObject<String, Void, Void> deferredObject =
+            new DeferredObject<String, Void, Void>();
+        Promise<String, Void, Void> promise = deferredObject.promise()
+            .except(new ExceptCallback() {
+                @Override
+                public void onException(Exception e) {
+                    assertEquals(THIS_IS_MY_RUNTIME_EXCEPTION, e.getMessage());
+                }
+            })
+            .always(new AlwaysCallback<String, Void>() {
+                @Override
+                public void onAlways(
+                    Promise.State state, String resolved, Void rejected
+                ) {
+                    throw new RuntimeException(THIS_IS_MY_RUNTIME_EXCEPTION);
+                }
+            });
+
+        deferredObject.reject(null);
+
+        try {
+            promise.waitSafely();
+            fail();
+        } catch (RuntimeException re) {
+        }
+    }
+
+    @Test
+    public void testProgressExceptionsAreCaught() throws InterruptedException {
+        final DeferredObject<String, Void, Void> deferredObject =
+            new DeferredObject<String, Void, Void>();
+        Promise<String, Void, Void> promise = deferredObject.promise()
+            .except(new ExceptCallback() {
+                @Override
+                public void onException(Exception e) {
+                    assertEquals(THIS_IS_MY_RUNTIME_EXCEPTION, e.getMessage());
+                }
+            })
+            .progress(new ProgressCallback<Void>() {
+                @Override
+                public void onProgress(Void progress) {
+                    throw new RuntimeException(THIS_IS_MY_RUNTIME_EXCEPTION);
+                }
+            });
+        deferredObject.notify(null);
+        deferredObject.reject(null);
+
+        try {
+            promise.waitSafely();
+            fail();
+        } catch (RuntimeException re) {
+        }
+    }
+
```

Comment:
Should we handle uncaught exception in the DeferredManager or in Promise callback?

E.g.,

```
dm.handleUncaughtException(e -> ...)
```

or

```
p.uncaughtException(e -> ...)